### ISSUE TITLE
[FEATURE REQ] Add Deferred Supplier APIs for Logging

### ISSUE THREAD

**Issue by alzimmermsft** [Azure/azure-sdk-for-java#19523]
Add `Supplier<String>` overloads for the logging methods in `ClientLogger`. These APIs will allow for better handling of cases where a log message may be dependent on conditional statements.

For example, without a `Supplier` based API this may need to be written.

```java
// By default there is two arguments but an additional third one may be added.
if (args.length >= 3) {
    logger.info("Param 1: {}, Param 2: {}, Param 3: {}", args);
} else {
    logger.info("Param 1: {}, Param 2: {}", args);
}
```

But when adding `Supplier` based APIs this is simplified to this.

```java
logger.info(() -> (args.length >= 3) 
    ? String.format("Param 1: %s, Param 2: %s, Param 3: %s", args);
    : String.format("Param 1: %s, Param 2: %s", args));
```

Additionally, given that this is a supplier, costly logic could be added as it won't be ran until the logger gets the supplier result, allowing for other logic to be simplified as follows. Though, if the costly logic was in calling `toString` on a format parameter this logic is already deferred, it is anything that requires non-`toString` formatting which is not.

```java
logger.info("A costly format requirement: {}", format(obj));
```

Is simplified to this.

```java
logger.info(() -> String.format("A costly format that is now deferred: %s", format(obj)));
```

**Comment by tozsvath** [comment]
Hello,

Can I take it? Looking for my first open-source contribution, for my long weekend.

**Comment by alzimmermsft** [comment]
Hi @tozsvath, I'll happily assign you the issue 😃 Thank you for taking interest in it.

**Comment by tozsvath** [comment]
Is it possible I don't have permission for the repo? I can't push my changes to my branch. I have tried to search in Microsoft Azure Projects Contribution Guidelines, but I got a 404 for that too.

**Comment by alzimmermsft** [comment]
Hi @tozsvath, I've updated the CONTRIBUTING.md to fix the link that 404s, thank you for notifying me of that.

For submitting the PR, are you attempting to create a branch on the main repo? Opening PRs should be available to everyone but only maintainers of the repository are able to create branches against the main repo. Everyone else submitting PRs should fork the repository, create their change branch, and then submitting that branch as the PR.

**Comment by alzimmermsft** [comment]
Hi @tozsvath, are you still blocked on being able to submit a PR?

**Comment by tozsvath** [comment]
Hi @alzimmermsft,

Yes the issue resolved, thanks for your help! Mr raised: https://github.com/Azure/azure-sdk-for-java/pull/20714

**Comment by tozsvath** [comment]
Is it possible for me to rerun the ci? I can see few weird errors

**Comment by alzimmermsft** [comment]
This was resolved by #20714
